utils/src/main/scala/za/co/absa/enceladus/utils/types/TypedStructField.scala

```diff
@@ -0,0 +1,231 @@
+import za.co.absa.enceladus.utils.schema.MetadataKeys
+import za.co.absa.enceladus.utils.time.DateTimePattern
+import za.co.absa.enceladus.utils.types.parsers._
+import za.co.absa.enceladus.utils.validation.ValidationIssue
+import za.co.absa.enceladus.utils.validation.field.{DateFieldValidator, DateTimeFieldValidator, FieldValidator, ScalarFieldValidator, TimestampFieldValidator}
+
+import scala.util.{Failure, Success, Try}
+
+sealed abstract class TypedStructField(structField: StructField) extends StructFieldEnhancements(structField) {
+  protected def convertString(string: String): Try[Any]
+
+  def validate():Seq[ValidationIssue]
+
+  def stringToTyped(string: String): Try[Option[Any]] = {
+    if (string == null) {
+      if (structField.nullable) {
+        Success(None)
+      } else {
+        Failure(new IllegalArgumentException(s"null is not a valid value for field '${structField.name}'"))
+      }
+    } else {
+      convertString(string).map(Some(_))
+    }
+  }
+
+  def ownDefaultValue: Try[Option[Option[Any]]] = {
+    if (hasMetadataKey(MetadataKeys.DefaultValue)) {
+      for {
+        defeaultValueString <- Try{structField.metadata.getString(MetadataKeys.DefaultValue)}
+        defaultValueTyped <- stringToTyped(defeaultValueString)
+      } yield Some(defaultValueTyped)
+    } else {
+      Success(None)
+    }
+  }
+
+  def defaultValueWithGlobal: Try[Option[Any]] = {
+    for {
+      localDefault <- ownDefaultValue
+      result <- localDefault match {
+        case Some(value) => Success(value)
+        case None => Defaults.getGlobalDefaultWithNull(dataType, nullable)
+      }
+    } yield result
+  }
+
+  def pattern: Try[Option[TypePattern]] = {
+    Success(None)
+  }
+
+  def name: String = structField.name
+  def nullable: Boolean = structField.nullable
+  def dataType: DataType = structField.dataType
+}
+
+object TypedStructField {
+  def apply(structField: StructField): TypedStructField = {
+    structField.dataType match {
+      case _: StringType =>StringTypeStructField(structField)
+      case _: BooleanType => BooleanTypeStructField(structField)
+      case _: ByteType => IntegralTypeStructField(structField)
+      case _: ShortType => IntegralTypeStructField(structField)
+      case _: IntegerType => IntegralTypeStructField(structField)
+      case _: LongType => IntegralTypeStructField(structField)
+      case _: FloatType => FractionalTypeStructField(structField)
+      case _: DoubleType => FractionalTypeStructField(structField)
+      case dt: DecimalType => DecimalTypeStructField(structField, dt)
+      case _: TimestampType => DateTimeTypeStructField(structField, TimestampFieldValidator)
+      case _: DateType => DateTimeTypeStructField(structField, DateFieldValidator)
+      case at: ArrayType => ArrayTypeStructField(structField, at)
+      case st: StructType => StructTypeStructField(structField, st)
+      case _ => GeneralTypeStructField(structField)
```

Comment:
1. Just the indentations
```scala
  def apply(structField: StructField): TypedStructField = {
    structField.dataType match {
      case _: StringType      => StringTypeStructField(structField)
      case _: BooleanType     => BooleanTypeStructField(structField)
      case _: ByteType        => IntegralTypeStructField(structField)
      case _: ShortType       => IntegralTypeStructField(structField)
      case _: IntegerType     => IntegralTypeStructField(structField)
      case _: LongType        => IntegralTypeStructField(structField)
      case _: FloatType       => FractionalTypeStructField(structField)
      case _: DoubleType      => FractionalTypeStructField(structField)
      case dt: DecimalType    => DecimalTypeStructField(structField, dt)
      case _: TimestampType   => DateTimeTypeStructField(structField, TimestampFieldValidator)
      case _: DateType        => DateTimeTypeStructField(structField, DateFieldValidator)
      case at: ArrayType      => ArrayTypeStructField(structField, at)
      case st: StructType     => StructTypeStructField(structField, st)
      case _                  => GeneralTypeStructField(structField)
    }
  }
```
2. I tried putting all integral together and looks supper ugly ...
3. By this time, we should already have our own dataType called integral and fractional.